### PR TITLE
Fix AirLoopHVACUnitarySystem FT issues with UnitarySystemPerformanceMultispeed

### LINKED ISSUE #5277: Forward Translator Issues for UnitarySystemPerformance:Multispeed
Issue overview
--------------
I am seeing a few issues with the forward translator for the UnitarySystemPerformance:Multispeed object. Some cause simulation failure, while others impact the intended operation of the system.

## Current Behavior
(1) The Heating (or Cooling) Speed Supply Air Flow Ratio X fields seem to appear only to the highest speed of cooling, rather than the highest speed of cooling or heating. This can result in speed levels being excluded from operation. The coil objects contain all appropriate speed data - the issue appears contained to the UnitarySystemPerformance:Multispeed object.

Example where Number of Speeds for Cooling > Number of Speeds for Heating
image

Example where Number of Speeds for Heating > Number of Speeds for Cooling
image

(2) When there are more cooling speed levels than heating speed levels, the Heating Speed Supply Air Flow Ratio X fields for the speed levels above the number of speeds for heating are left blank, causing simulation failure. According to the documentation, these fields for speeds higher than the "number of speeds" should still be articulated. (See image above where Number of Speeds for Cooling > Number of Speeds for Heating).     

## Expected Behavior
The number of Heating (or Cooling) Speed Supply Air Flow Ratio X fields should match the higher of the Number of Speeds for Heating and Number of Speeds for Cooling. And any speed level above the Number of Speeds for Heating/Cooling should be populated with a 1.

## Steps to Reproduce
1. Define an OpenStudio that uses multispeed heating and cooling coils in a unitary system. One instance should have more cooling stages, and the other should have more heating stages.
2. Run the model or forward translator to get the resulting IDF.
3. Examine the UnitarySystemPerformance:Multispeed objects to see the described behavior.

Example OSM and resulting IDF when Num Speeds Heating > Num Speeds Cooling:
[in_MoreHeatingSpeeds.osm.txt](https://github.com/user-attachments/files/17479088/in_MoreHeatingSpeeds.osm.txt)
[in_MoreHeatingSpeeds.idf.txt](https://github.com/user-attachments/files/17479089/in_MoreHeatingSpeeds.idf.txt)

Example OSM and resulting IDF when Num Speeds Cooling > Num Speeds Heating:
[in_MoreCoolingSpeeds.osm.txt](https://github.com/user-attachments/files/17479943/in_MoreCoolingSpeeds.osm.txt)
[in_MoreCoolingSpeeds.idf.txt](https://github.com/user-attachments/files/17479944/in_MoreCoolingSpeeds.idf.txt)

## Possible Solution
I don't know if this is the issue, but the code below shows the maxStages variable getting set first by heating and then the same variable name set by cooling before being used in the for loop that follows. Because cooling is set second, that could explain why the number of  Speed Supply Air Flow Ratio X fields seems to follow the number of speeds for cooling.

https://github.com/NREL/OpenStudio/blob/develop/src/energyplus/ForwardTranslator/ForwardTranslateAirLoopHVACUnitarySystem.cpp#L483-L532

### Environment

Some additional details about your environment for this issue (if relevant):
 - Platform (Operating system, version): Windows
 - Version of OpenStudio (if using an intermediate build, include SHA): OpenStudio 3.8.0

### Context
These errors limit a user's ability to model multispeed systems with different numbers of heating/cooling speeds. Getting this to work properly requires edits to the IDF, which is not realistic for many applications. We are using these objects for modeling measure scenarios in ComStock.

### PR DESCRIPTION
**Pull request by joseph-robertson**
Pull request overview
---------------------

 - Fixes #5277

### Pull Request Author

 - [ ] Model API Changes / Additions
 - [ ] Any new or modified fields have been implemented in the EnergyPlus ForwardTranslator (and ReverseTranslator as appropriate)
 - [ ] Model API methods are tested (in `src/model/test`)
 - [ ] EnergyPlus ForwardTranslator Tests (in `src/energyplus/Test`)
 - [ ] If a new object or method, added a test in NREL/OpenStudio-resources: Add Link
 - [ ] If needed, added VersionTranslation rules for the objects (`src/osversion/VersionTranslator.cpp`)
 - [ ] Verified that C# bindings built fine on Windows, partial classes used as needed, etc.
 - [ ] All new and existing tests passes
 - [ ] If methods have been deprecated, update rest of code to use the new methods

**Labels:**

 - [ ] If change to an IDD file, add the label `IDDChange`
 - [ ] If breaking existing API, add the label `APIChange`
 - [ ] If deemed ready, add label `Pull Request - Ready for CI` so that CI builds your PR

### Review Checklist

This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] Code Style, strip trailing whitespace, etc.
 - [ ] All related changes have been implemented: model changes, model tests, FT changes, FT tests, VersionTranslation, OS App
 - [ ] Labeling is ok
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified

### CODE REVIEW

File: src/energyplus/ForwardTranslator/ForwardTranslateAirLoopHVACUnitarySystem.cpp [hunk 1]
```diff
@@ -566,7 +567,7 @@ namespace energyplus {
           } else {
             extensible.setString(UnitarySystemPerformance_MultispeedExtensibleFields::HeatingSpeedSupplyAirFlowRatio, "Autosize");
           }
-        } else if (i < 2) {
```
Comment (joseph-robertson) [inline src/energyplus/ForwardTranslator/ForwardTranslateAirLoopHVACUnitarySystem.cpp:569]: Not sure the purpose of the "< 2"...?

File: src/energyplus/ForwardTranslator/ForwardTranslateAirLoopHVACUnitarySystem.cpp [hunk 2]
```diff
@@ -482,36 +483,36 @@ namespace energyplus {
...
-        _unitarySystemPerformance.setInt(UnitarySystemPerformance_MultispeedFields::NumberofSpeedsforHeating, maxStages);
+        maxHeatingStages = waterToAirHeatingStages.size();
+        _unitarySystemPerformance.setInt(UnitarySystemPerformance_MultispeedFields::NumberofSpeedsforHeating, maxHeatingStages);
       } else {
         _unitarySystemPerformance.setInt(UnitarySystemPerformance_MultispeedFields::NumberofSpeedsforHeating, 1);
       }
 
       if (multispeedDXCooling) {
         coolingStages = multispeedDXCooling->stages();
-        maxStages = coolingStages.size();
```
Comment (joseph-robertson) [inline src/energyplus/ForwardTranslator/ForwardTranslateAirLoopHVACUnitarySystem.cpp:505]: This was overwriting `maxStages` set by heating speeds.

File: src/energyplus/Test/UnitarySystemPerformanceMultispeed_GTest.cpp [hunk 3]
```diff
@@ -0,0 +1,435 @@
...
+
+    IdfExtensibleGroup eg2 = egs[1];
+    ASSERT_FALSE(eg2.getDouble(UnitarySystemPerformance_MultispeedExtensibleFields::HeatingSpeedSupplyAirFlowRatio));
+    EXPECT_EQ("autosize", eg2.getString(UnitarySystemPerformance_MultispeedExtensibleFields::HeatingSpeedSupplyAirFlowRatio).get());
+    ASSERT_TRUE(eg2.getDouble(UnitarySystemPerformance_MultispeedExtensibleFields::CoolingSpeedSupplyAirFlowRatio));
+    EXPECT_EQ(0.3, eg2.getDouble(UnitarySystemPerformance_MultispeedExtensibleFields::CoolingSpeedSupplyAirFlowRatio).get());
+  }
+}
+
+TEST_F(EnergyPlusFixture, ForwardTranslator_UnitarySystemPerformanceMultispeed_AirLoopHVACUnitarySystem) {
```
Comment (joseph-robertson) [inline src/energyplus/Test/UnitarySystemPerformanceMultispeed_GTest.cpp:239]: New test for translation of `UnitarySystemPerformanceMultispeed` using `AirLoopHVACUnitarySystem`.

**Review by jmarrec — APPROVED** [review]

File: src/energyplus/Test/UnitarySystemPerformanceMultispeed_GTest.cpp [hunk 4]
```diff
@@ -0,0 +1,471 @@
...
+      EXPECT_TRUE(htgstage.setRatedAirFlowRate(i));
+      EXPECT_TRUE(htgcoil.addStage(htgstage));
+    }
+
+    CoilCoolingDXMultiSpeed clgcoil(m);
+    for (int i = 1; i <= maxCoolingStages; ++i) {
+      CoilCoolingDXMultiSpeedStageData clgstage(m);
+      EXPECT_TRUE(clgstage.setRatedAirFlowRate(i + maxHeatingStages));
+      EXPECT_TRUE(clgcoil.addStage(clgstage));
+    }
```
Comment (jmarrec) [inline src/energyplus/Test/UnitarySystemPerformanceMultispeed_GTest.cpp:404]: Make this test case use 2 speed for heating and 4 for cooling.

File: src/energyplus/Test/UnitarySystemPerformanceMultispeed_GTest.cpp [hunk 5]
```diff
@@ -156,19 +156,22 @@ TEST_F(EnergyPlusFixture, ForwardTranslator_UnitarySystemPerformanceMultispeed_C
...
+
+    {
+      const IdfExtensibleGroup& eg = egs[0];
+      ASSERT_TRUE(eg.getDouble(UnitarySystemPerformance_MultispeedExtensibleFields::HeatingSpeedSupplyAirFlowRatio));
+      EXPECT_EQ(0.1, eg.getDouble(UnitarySystemPerformance_MultispeedExtensibleFields::HeatingSpeedSupplyAirFlowRatio).get());
+      ASSERT_TRUE(eg.getDouble(UnitarySystemPerformance_MultispeedExtensibleFields::CoolingSpeedSupplyAirFlowRatio));
+      EXPECT_EQ(0.2, eg.getDouble(UnitarySystemPerformance_MultispeedExtensibleFields::CoolingSpeedSupplyAirFlowRatio).get());
+    }
+    {
+      const IdfExtensibleGroup& eg = egs[1];
```
Comment (jmarrec) [inline src/energyplus/Test/UnitarySystemPerformanceMultispeed_GTest.cpp:169]: scoping

**Review by jmarrec — COMMENTED** [review]
@joseph-robertson I made some nitpick changes and tested the last one with 2 heating and 4 cooling stages. Take a quick look. Otherwise this is looks good. Thanks again for the thorough testing!

**Pull request merged**